moz/l10n/bin/build.py

```diff
@@ -67,61 +70,144 @@ def cli() -> None:
-                            or section.id + entry.id in source_ids
-                        ]
-                        msg_delta += len(section.entries)
-                    msg = f"filter {rel_path}"
-                    log.info(f"{msg} ({msg_delta})" if msg_delta != 0 else msg)
-                    with open(tgt_path, "w") as file:
-                        for line in serialize_resource(res, trim_comments=True):
-                            file.write(line)
-                elif l10n_base != l10n_target:
+            if source:
+                msg_delta = write_target_file(rel_path, source, l10n_path, tgt_path)
+                if msg_delta < 0:
+                    msg_data[locale][0] -= msg_delta
+                elif msg_delta > 0:
+                    msg_data[locale][1] += msg_delta
+                else:
+                    msg_data[locale]
+            elif exists(l10n_path):
+                if l10n_base != l10n_target:
                     log.info(f"copy {rel_path}")
                     copyfile(l10n_path, tgt_path)
                 else:
                     log.info(f"skip {rel_path}")
-            elif source_ids:
-                log.info(f"create empty {rel_path}")
-                open(tgt_path, "a").close()
             else:
                 log.info(f"skip {rel_path}")
 
+    log.info("----")
+    for locale, (ftl_missing, l10n_fallback) in sorted(
+        msg_data.items(), key=lambda d: d[0]
+    ):
+        log.info(f"{locale}:")
+        log.info(f"  missing_ftl {ftl_missing:>6}")
+        log.info(f"  fallback    {l10n_fallback:>6}")
+
+
+def write_target_file(
+    name: str,
+    source_res: Resource[Message, str],
+    l10n_path: str,
+    tgt_path: str,
+) -> int:
+    l10n_res = parse_resource(l10n_path) if exists(l10n_path) else None
+
+    if source_res.format == Format.fluent:
+        # Fluent uses per-message fallback at runtime, allowing resources to be incomplete.
+        # Therefore, build result by filtering out any content not in the source.
+        if l10n_res is None:
+            msg_delta = -sum(
+                sum(1 for entry in section.entries if isinstance(entry, Entry))
+                for section in source_res.sections
+            )
+            log.info(f"create empty {name} ({msg_delta})")
+            open(tgt_path, "a").close()
+            return msg_delta
+
+        source_map = _message_map(name, source_res)
+        msg_delta = 0
+        for section in l10n_res.sections:
+            msg_delta -= len(section.entries)
+            section.entries = [
+                entry
+                for entry in section.entries
+                if isinstance(entry, Entry) and section.id + entry.id in source_map
+            ]
+            msg_delta += len(section.entries)
+        msg = f"filter {name}"
+        log.info(f"{msg} ({msg_delta})" if msg_delta != 0 else msg)
+        with open(tgt_path, "w") as file:
+            for line in serialize_resource(l10n_res, trim_comments=True):
```

Comment:
I think you lost me here: you're trimming comments here, but where are you removing obsolete strings from the localized resource?

`l10n_res` hasn't been changed since `parse_resource(l10n_path) if exists(l10n_path) else None` (if the l10n resources is not empty).

Reply:
The filtering of `l10n_res` contents was done here in the assignment to `section.entries` on line 145, but I refactored the Fluent handling to also be based on iterating the source, so this is out of date.

Reply:
Completely missed that `for section in l10n_res.sections` would affect `l10n_res` as an object 🤦🏼